File: Data Structures & Algorithms/products-of-array-discluding-self/submission-3.swift
class Solution {
    func productExceptSelf(_ nums: [Int]) -> [Int] {

        var product = 1
        var zeroCount = 0

        for i in 0..<nums.count{
            if nums[i] == 0 { 
                zeroCount += 1
            }else{
                product = product * nums[i]
            }

        }

        var result: [Int] = []
        
        for num in nums{
           

            if zeroCount > 1{
                result.append(0)

            } else if zeroCount == 1{
                result.append(num == 0 ? product : 0)
            }else{
                result.append(product/num)
            }   
           
           
        }

        return result



    }
}

File: Data Structures & Algorithms/house-robber-ii/submission-3.java
class Solution {
    public int rob(int[] nums) {
        int[]dp = new int[nums.length];

         if(nums.length == 1){
            return nums[0];
        }

        

        dp[0] = nums[0];
        dp[1] = Math.max(nums[1],nums[0]);

        if(nums.length == 2){
            return dp[1];
        }



       return Math.max(rob(0,nums.length-1, nums,dp),rob(1,nums.length, nums,dp));
        
    }

    public int rob(int st,int e,int[] nums,int[] dp){

       

        dp[st] = nums[st];
        dp[st+1] = Math.max(nums[st+1],nums[st]);

        for(int i=st+2;i<e;i++){
            dp[i] = Math.max(dp[i-2]+nums[i],dp[i-1]);
        }

        return dp[e-1];
    }
}
